src/api/ping.rs

```diff
@@ -0,0 +1,59 @@
+use actix_identity::Identity;
+use actix_web::{web, HttpResponse};
+use diesel::{insert_into, ExpressionMethods, PgJsonbExpressionMethods, RunQueryDsl};
+use serde::Deserialize;
+use serde_json::{json, Value};
+
+use crate::db::{self, model::ActivityPingInsert, schema::activity_pings, Pool};
+
+use super::error::ApiError;
+
+#[derive(Deserialize)]
+pub struct PingQuery {
+    pub offline: Option<bool>,
+}
+
+pub async fn ping(
+    form: web::Form<PingQuery>,
+    id: Option<Identity>,
+    pool: web::Data<Pool>,
+) -> Result<HttpResponse, ApiError> {
+    match id {
+        Some(id) => {
+            let mut conn_pool = pool.get()?;
+            let user = db::users::get_user(&mut conn_pool, id.id().unwrap());
+            match user {
+                Ok(found) => {
+                    let mut activity_data = json!({
+                        "subscription_type": found.get_subscription_type()
+                    });
+
+                    if form.offline.unwrap_or(false) {
+                        // careful: we don't include the offline key
```

Comment:
Is there any reason not to make the ActivityPing an array that added to with each ping? This latching on behaviour is fine but perhaps there is soemthig to be learned from users mashing the on off button?

Reply:
You're right, it would certainly give us more context, though I think it's a little difficult to interpret because while multiple items in the array could be a user toggling on/off (though I haven't fully worked out if we'll send pings when this happens, it's not completely necessarily for working out DAU/MAU, at least in the "off" direction), it could also be a result of multiple clients being logged in for a single user, or even a user logging out then logging in again (or clearing their localstorage through other means).

What it *could* tell us in the future, if we extend these pings to other features, is if two features are being used together on the same client/session (dependent on exactly how we implement those pings client side).

Though I wonder if it'll make things a bit more complicated on the querying for the DAU/MAU side? I guess it's just one extra array function on top of what would already exist.

You're the metricsmeister so I'll defer decision making on this to you ;)